keras_nlp/samplers/beam_sampler.py

```diff
@@ -0,0 +1,215 @@
+# Copyright 2023 The KerasNLP Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Beam Sampler."""
+
+import tensorflow as tf
+from tensorflow import keras
+
+from keras_nlp.samplers.sampler import Sampler
+from keras_nlp.samplers.sampler import base_sampler_args_docstring
+from keras_nlp.samplers.sampler import call_args_docstring
+from keras_nlp.samplers.sampler import sample_args_docstring
+from keras_nlp.utils.python_utils import format_docstring
+
+
+@format_docstring(
+    base_sampler_args=base_sampler_args_docstring, call_args=call_args_docstring
+)
+@keras.utils.register_keras_serializable(package="keras_nlp")
+class BeamSampler(Sampler):
+    """Beam Sampler class.
+
+    This sampler implements beam search algorithm. At each time-step, beam
+    search keeps the beams (sequences) of the top `num_beams` highest
+    accumulated probabilities, and uses each one of the beams to predict
+    candidate next tokens.
+
+    Args:
+        num_beams: int. The number of beams that should be kept at each
+            time-step. `num_beams` should be strictly positive.
+        {{base_sampler_args}}
+
+    Call Args:
+        {{call_args}}
+
+    Examples:
+    ```python
+    BATCH_SIZE = 8
```

Comment:
@chenmoneygithub I really don't think these arg blocks match the rest of the library's style. I don't think there's a perfect answer but I'd prefer it not to be obvious who wrote what 🥗

Reply:
Thanks! I checked our code base, it appears we have a few code having this style:
- maskedLmhead: [link](https://github.com/keras-team/keras-nlp/blob/master/keras_nlp/layers/masked_lm_head.py#L71)
- PositionEmbedding: [link](https://github.com/keras-team/keras-nlp/blob/master/keras_nlp/layers/position_embedding.py#L54)
- SinePositionEmbedding: [link](https://github.com/keras-team/keras-nlp/blob/master/keras_nlp/layers/position_embedding.py#L54)
- Generation utils: [link](https://github.com/keras-team/keras-nlp/blob/master/keras_nlp/utils/text_generation.py#L131)

I feel like giving these numbers an explicit meaning makes the code longer but kinda easier to understand.

Reply:
It's a totally valid opinion @chenmoneygithub but it's also important for the code to have a unified style rather than each contributor producing different looking code. If there's an example that's unclear without named params I'm open to trying something different but otherwise I'm hoping we can compromise!

Reply:
Taking steps towards a more unified style (and then reflecting that in our style guide) sgtm. What are the main places this differs, besides the constants at the top?

Reply:
Basically what I asked for in https://github.com/keras-team/keras-nlp/issues/658 was the current thought. A bit less script-like and a little more "drop this line in colab and see what we're talking about".

Reply:
I could see us replacing the "model" with something like `tf.random.uniform(shape, minval=-1, maxval=1)`. It is kind of weird to me that we show a whole model, that is trainable but randomly initialized (so results will be random anyway), and not even sequence aware so would never really perform even if your trained it. For a new user this seems a bit of a red herring.

Would be more concise to do something like:
```python
    def token_probability_fn(inputs, mask):
        return tf.random.uniform(...) # Replace with a real model!
```

Reply:
I would like to keep the model part so that the example is closer to real use cases. 

I will unify the docstring to move those hypers inline.